init.py

```diff
@@ -180,24 +181,89 @@ def sierpinski(t, l, r, deg, preds, g):
             add_edges(l,t,preds,g)
             add_edges(t,r,preds,g)
             add_edges(r,l,preds,g)
-            
+
     def add_edges(u,v,preds,g):
         ''' Adds edges between vertices u and v for all predicates. '''
         for p in preds:
             g += [[u,p,v]]
             g += [[v,p,u]]
-            
+
     def _idgen():
         ''' Generates integer identifiers for vertices. '''
         c = 4
         while True:
             yield c
             c += 1
-            
+
     ids = _idgen()
     graph = []
-    sierpinski(1,2,3,degree,predicates, graph)   
-    with open(os.path.join(target_dir, 'sierpinskigraph_%s.txt'%(degree)), 'w') as out_file:
+    sierpinski(1,2,3,degree,predicates, graph)
+    filename = 'sierpinskigraph_%d_%s.txt' % (degree,''.join(predicates))
+    _to_file(target_dir, filename, graph)
+
+def gen_kuijpers_random(target_dir, vertices, k, labels='A B C D'):
+    '''Generate random graphs as defined on J. Kuijpers's paper "An
+    Experimental Study of Context-Free Path Query Evaluation Methods".
+    '''
+    n = vertices
+    L = labels.split(' ')
+    assert n >= k, 'It is required that vertices >= k'
+
+    def label(L):
+        '''Randomly choose a label in `L`'''
+        return random.choice(L)
+
+    def add(D,i,j,I,var):
+        '''Add an edge to `D`.'''
+        l = label(L)
+        D += [[i,l,j]]
+        D += [[j,l+"R",i]]
+        if i == len(I):
+            I += [0]
+        I[j] += 1
+        I[i] += 1
+        var['sum'] += 2
+
+    def p(i,I,sumI):
+        '''Calculate the probability of adding an edge to vertex `i`.'''
+        if sumI == 0:
+            return 1
+        return (I[i]) / sumI
+
+    D = [] # the graph
+    I = [] # vertices' degree
+    var = {'sum' : 0} # sum of vertices' degree
+    for i in range(k):
+        for j in range(k):
+            if i != j:
+                l = label(L)
+                D += [[i,l,j]]
+                D += [[j,l+"R",i]]
+    I = [(k-1)*2]*k
+    var['sum'] = (k-1)*2*k
+    for i in range(k,n):
+        chosenset = set()
+        chosensum = 0.0
+        for _ in range(k):
+            r = random.uniform(0,1-chosensum)
```

Comment:
There is a built in function `random.choices`, which requires some container. Also it has a additional argument of array of probabilities. I think the code will become more clear if we express the choice of k vertices using this function.

Reply:
Yes!
I didn't know this function, that's why I've implemented the whole thing.
Your implementation looks better, you may regret mine. Keep only the _to_file function, if you want, I believe it can be useful.

Reply:
`_to_function` is really useful, we should to merge it.